feat(eslint-plugin): add `no-duplicate-type-constituents` rule

sajikix opened the pull request:
<!--
👋 Hi, thanks for sending a PR to typescript-eslint! 💖
Please fill out all fields below and make sure each item is true and [x] checked.
Otherwise we may not be able to review your PR.
-->

## PR Checklist

- [x] Addresses an existing open issue: fixes #2738
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [CONTRIBUTING.md](https://github.com/typescript-eslint/typescript-eslint/blob/main/CONTRIBUTING.md) were taken

## Overview

Add new rule : `no-duplicate-type-union-intersection-members`.

This rule disallows duplicate union or intersection type members. The autofixer will remove all duplicates, keeping the first one.

I implemented it so that only notational duplicates are checked. 
It means that members with the same value but different names are not marked duplicates. 
```ts
type A = { a: string; b: number };
type B = { a: string; b: number };

type AorB = A | B; // Type A and B are the same type, but are not considered duplicates.
```

In the case of object types, it judged as duplicates only when the order of key and value are exactly the same.

```ts
// ❌ berfore remove duplicates
type T =
  | { a: string; b: number }
  | { a: string; b: number } // considered duplicates with the first one. 
  | { b: number; a: string }; // not considered duplicates with the first one. 
// ↓ 
// ✅ after remove duplicates
type T =
  | { a: string; b: number }
  | { b: number; a: string };
```

What has considered a duplicate in `object type` is a matter of debate.  I would love to hear the reviewers' opinions and would like to improve it with the reviewers' input.

JoshuaKGoldberg commented on packages/eslint-plugin/src/rules/no-duplicate-type-union-intersection-members.ts:
```diff
@@ -0,0 +1,122 @@
+    }
+    return {
+      ...(checkIntersections && {
+        TSIntersectionType(node): void {
+          checkDuplicate(node);
+        },
```
Nit: for both of these, a teeny bit less code?

```suggestion
        TSIntersectionType: checkDuplicate
```

sajikix replied:
Yeah, This part could be written more simply. 👍

JoshuaKGoldberg commented on packages/eslint-plugin/src/rules/no-duplicate-type-union-intersection-members.ts:
```diff
@@ -0,0 +1,122 @@
+        const result = [...uniqueMemberTexts].join(
+          node.type === AST_NODE_TYPES.TSIntersectionType ? ' & ' : ' | ',
+        );
+        return fixer.replaceText(node, result);
```
Instead of doing a full replacement of the node, this could `fixer.remove` to remove any duplicate nodes. A couple benefits:

* Comments aren't so aggressively trimmed out
* User formatting is mangled much less

JoshuaKGoldberg commented on packages/eslint-plugin/src/rules/no-duplicate-type-union-intersection-members.ts:
```diff
@@ -0,0 +1,122 @@
+  defaultOptions: [
+    {
+      checkIntersections: true,
+      checkUnions: true,
```
IME it's confusing for folks to have the default value for something be truthy. How about switching these to `ignoreIntersections` & `ignoreUnions`?

sajikix replied:
Exactly! 
This rule check both intersection and union duplication by default, so I think `ignore` option would be clearer!

JoshuaKGoldberg commented on packages/eslint-plugin/docs/rules/no-duplicate-type-union-intersection-members.md:
```diff
@@ -0,0 +1,49 @@
+
+> 🛑 This file is source code, not the primary documentation location! 🛑
+>
+> See **https://typescript-eslint.io/rules/no-duplicate-type-union-intersection-members** for documentation.
```
"Members" typically refers to class/object members. The technical term here is _constituents_, I think. To match with https://typescript-eslint.io/rules/no-redundant-type-constituents, how about...

```suggestion
> See **https://typescript-eslint.io/rules/no-duplicate-type-constituents** for documentation.
```

Aside: I had a jolt of panic looking at this issue because at first I misinterpreted it to be a dup of that rule. But these are different, phew!

sajikix replied:
> "Members" typically refers to class/object members. The technical term here is constituents, I think. 

Thanks for the useful info! (I'm not yet familiar with the nuances of English technical terms.... So it was very helpful 👍 ).
I changed the rule names and file names as suggested.

JoshuaKGoldberg commented on packages/eslint-plugin/src/rules/no-duplicate-type-union-intersection-members.ts:
```diff
@@ -0,0 +1,122 @@
+      const source = node.types.map(type => {
+        return {
+          node: type,
+          text: sourceCode.getText(type),
```
We generally try not to use `sourceCode.getText` or similar in rules if at all possible. It's less reliable than comparing AST and/or types - for example, it'd be very confusing to users why adding a comment inside a type stops this rule from complaining on it. I'm not looking forward to fielding those bug reports 😄.

Does using the type checker and checking types for `===` work? I.e. building a `Set` of types for each union/intersection, and complaining for each type that's the same as a previous one?

Obligatory grumble that https://github.com/microsoft/TypeScript/issues/9879 isn't implemented...

JoshuaKGoldberg reviewed (changes requested):
A great start, thanks for sending this in! 💓 

Left some requests for changes - lmk if any of that doesn't make sense!

sajikix commented:
Sorry for the wait since I got your review!

There have been two major changes since previous review.
1. change duplicate check to recursive ASTNode comparison instead of string comparison from `sourceCode.getText`.
2. use `removeRange` instead of `replaceText` in the fix function.

#### 1. use recursive ASTNode comparison
As you reviewed, the previous implementation compared pure code strings, so there was a bug that prevented comparison if there were comments, etc. in the union/intersction type.
To resolve this, I changed to compare ASTNode recursively. Specifically, this rule is looking for an exact match for each ASTNode excluding  `range`, `loc`, and `parent` properties. (I first tried to write a comparison process for each AstTypeNode, but there are so many cases that I am not confident I could cover them... 😭 )
I believe this method can be used to compare union/intersection members, but do you have any concerns? 

#### 2. use removeRange()
As you reviewed, we are using removeRange instead of replaceText in the fix function. Specifically, I set the removeRange as follows.

```ts
type T = A | A | B;
//        ^   ^
//     start  end

// start : end loc of the member before the duplicated member
// end : end loc of the duplicated member

// ↓ Deleting this range will result in ....
type T = A | B;

```

This is a relatively large change, but I would appreciate your review. 🙏

JoshuaKGoldberg commented on packages/eslint-plugin/src/rules/no-duplicate-type-constituents.ts:
```diff
@@ -0,0 +1,221 @@
+      }, []);
+
+      const hasComments = node.types.some(type => {
+        const count =
+          sourceCode.getCommentsBefore(type).length +
+          sourceCode.getCommentsAfter(type).length;
+        return count > 0;
```
Nit: if `getCommentsBefore` has entries, we don't need to call `getCommentsAfter`.

```suggestion
        return (
          sourceCode.getCommentsBefore(type).length ||
          sourceCode.getCommentsAfter(type).length
        );
```

JoshuaKGoldberg commented on packages/eslint-plugin/src/rules/no-duplicate-type-constituents.ts:
```diff
@@ -0,0 +1,221 @@
+  expectedTypeNode: TSESTree.TypeNode,
+): boolean => {
+  switch (expectedTypeNode.type) {
+    // These types should never occur as constituents of a union/intersection
+    case AST_NODE_TYPES.TSAbstractKeyword:
+    case AST_NODE_TYPES.TSAsyncKeyword:
+    case AST_NODE_TYPES.TSDeclareKeyword:
+    case AST_NODE_TYPES.TSExportKeyword:
+    case AST_NODE_TYPES.TSNamedTupleMember:
+    case AST_NODE_TYPES.TSOptionalType:
+    case AST_NODE_TYPES.TSPrivateKeyword:
+    case AST_NODE_TYPES.TSProtectedKeyword:
+    case AST_NODE_TYPES.TSPublicKeyword:
+    case AST_NODE_TYPES.TSReadonlyKeyword:
+    case AST_NODE_TYPES.TSRestType:
+    case AST_NODE_TYPES.TSStaticKeyword:
+    case AST_NODE_TYPES.TSTypePredicate:
+      /* istanbul ignore next */
+      throw new Error(`Unexpected Type ${expectedTypeNode.type}`);
```
If they should never occur, then we don't need to handle them. 😄 

```suggestion
```

JoshuaKGoldberg commented on packages/eslint-plugin/src/rules/no-duplicate-type-constituents.ts:
```diff
@@ -0,0 +1,221 @@
+      throw new Error(`Unexpected Type ${expectedTypeNode.type}`);
+
+    default:
+      if (nodeTypesAreEqual(actualTypeNode, expectedTypeNode)) {
+        return astNodesAreEquals(actualTypeNode, expectedTypeNode);
+      }
+      return false;
```
Nit: expressing the intent of the code a little bit more... what do you think?

```suggestion
      return (
        nodeTypesAreEqual(actualTypeNode, expectedTypeNode) &&
        astNodesAreEquals(actualTypeNode, expectedTypeNode)
      );
```

(this is definitely not a blocking comment if you prefer the way it is now!)

JoshuaKGoldberg commented on packages/eslint-plugin/src/rules/no-duplicate-type-constituents.ts:
```diff
@@ -0,0 +1,221 @@
+    }
+  }
+  return false;
+};
```
> ... changed to compare ASTNode recursively. Specifically, this rule is looking for an exact match for each ASTNode excluding `range`, `loc`, and `parent` properties ...

This won't work in all cases 😞. Types might be effectively the same but have slightly different AST representations. One of the test cases as-is demonstrates this:

```ts
type T = "1" | '1';
```

Another example is union types with different orders:

```ts
type T = (number | string)[] | (string | number)[];
```

I think this rule is going to have to use the type checker. Users are going to be rather disappointed that this (simplified from your example) doesn't give an error:

```ts
type A = string;
type B = string;
type C = A | B;
```

JoshuaKGoldberg commented on packages/eslint-plugin/src/rules/no-duplicate-type-constituents.ts:
```diff
@@ -0,0 +1,221 @@
+        return [...acc, cur];
+      }, []);
+
+      const hasComments = node.types.some(type => {
```
Users really, really dislike it when aspects of code unrelated to a rule change how that rule works. It'd be very confusing for the rule to only provide an autofixer when code _isn't_ commented. That'd feel like a bug.

My suggestion from before stands: use node-based fixer methods such as `fixer.remove`. That way comments are handled by ESLint the same way between rules. The `sourceCode` object has methods such as `getTokenAfter` if you need to also remove tokens like `&`.

JoshuaKGoldberg reviewed (changes requested):
Progress! I think this is getting closer - but nailing down edge cases (i.e. switching to `===` type equality) is going to be key.

JoshuaKGoldberg commented on packages/eslint-plugin/src/rules/no-duplicate-type-constituents.ts:
```diff
@@ -0,0 +1,221 @@
+  return actualTypeNode.type === expectedTypeNode.type;
+};
+
+const constituentsAreEqual = (
```
@sajikix I just realized I might not have very clear in my suggestion about using the type checker. I'm essentially proposing that we check whether the `ts.Type` for a constituent node is `===` any other previous constituent node.

```ts
const parserServices = util.getParserServices(context);
const checker = parserServices.program.getTypeChecker();

const type = checker.getTypeAtLocation(parserServices.parserServices.esTreeNodeToTSNodeMap.get(node));
```

It's a little silly to use `===`, but we're guaranteed[^1] that the TypeScript type checking APIs will use the same type object from two nodes that have the same type.

[^1]: I don't know that this has been explicitly documented anywhere, but a lot of TypeScript tools work on this assumption. It's now one of those things that TypeScript just has to do.

sajikix commented:
Thanks again for the review.
There were a few things I didn't understand (especially around Type Checker), and I was implementing them in a misguided way... 😢

### Comparison between constituents
As you advised, I implemented it by comparing Type Objects (using Set Object). In many cases, this comparison worked well and was smarter than a simple AST Type Node comparison. 

However, in the following cases, this implementation cannot detect that the types are the same.
```ts
type T = { a: string } | { a: string };
```
Users would want to count the above example as a duplicate. In the above example, only `id` property of TypeObject is different.
```
// Difference of type in the above code
// left constituent
{
  flags: 524288,
  id: 88,
  objectFlags: 16,
  symbol: SymbolObject {
    flags: 2048,
    escapedName: '__type',
    declarations: [ [NodeObject] ],
    members: Map(2) { 'a' => [SymbolObject], 'b' => [SymbolObject] },
    id: 11
  },
 ...[snip]
}
// right constituent
{
  flags: 524288,
  id: 89,
  objectFlags: 16,
  symbol: SymbolObject {
    flags: 2048,
    escapedName: '__type',
    declarations: [ [NodeObject] ],
    members: Map(2) { 'a' => [SymbolObject], 'b' => [SymbolObject] },
    id: 12
  },
 ...[snip]
}
```

 I suppose it is possible to ignore the `id` property and compare in such cases, but should I?


### use AST-based fixer method and bracket deletion
I have modified it to only use `fixer.remove()` to remove Node and Token.
Thus, there is no longer a need to change the behavior of the fixer depending on comments. (I had also received comments on the comment detection logic, but removed them...)

However, I have two concerns about these changes. 

1. space remains
Spaces between TOKENS are no longer removed. These spaces are not considered for deletion in this rule, right? (i.e., leave it to the lint rule or formatter set by user)

2. bracket deletion
I also included the following bracket deletion process in the implementation.
```ts
type T = (A) | (A)
// ↓
type T = A
```
However, I am wondering if this kind of bracket deletion process should also be in the scope of this rule. Generally, unnecessary brackets are removed by other linter-rules and formatter. However, if only the brackets are left as constituents of union / intersection, it will result in a TypeScript parsing error.

```ts
// This is thoros error
type T = "A" | ()
              ^^^
'Function type notation must be parenthesized when used in a union type.'
```

JoshuaKGoldberg commented:
> Comparison between constituents

Grr. I suppose it makes sense TypeScript would consider those to be different types. Having to know whether any arbitrary object literal is the same type as a previously existing type could be complex.

I suppose we'll have to go with _two_ heuristics to know whether two types are equivalent:

1. Recursively checking their AST to check node equality
2. type checker `===`

We'd want to do them in that order because asking the type checker for anything can be slow. But if the ASTs aren't the same, we'd still need to check the types for `===`

> space remains

We don't care 😄. If users want to have formatting enforced, they should use a tool such as Prettier. 

Whitespace code tends to be very verbose and difficult to maintain. It's better to have a slightly worse formatted output from an autofix than to have to maintain many -often hundreds- of lines of code.

> bracket deletion

Nit: I think you're referring to _parenthesis_ removal? If so: we only care about them when they cause syntax parsing errors, as you've described with the `"A" | ()`. In that case they'll need to be removed, yes.

Fixing from `(A) | (A)` to either `(A)` or `A` is equally fine output to us. There may even be users who intentionally wrapped the `A`s with parenthesis! Whichever takes less code to implement (I would assume not changing parenthesis) is best.

sajikix commented:
Thanks so much for your multiple reviews! 🙏

I have made some rule modifications and added some test cases as you advised.

### Comparisons between constituents.
I would like to use both a combination of recursive comparison of AST nodes and type object comparison. This would also allow for object literal comparisons. (It doesn't recognize that they are of the same type if the order of the properties is different...but I think it is still useful enough. )
For AST node comparisons, I am using the one I created earlier in this PR.

### Parenthesis removal
I have implemented this in a way that does not change the user's code as much as possible. I also added a test for this behavior.

JoshuaKGoldberg commented on packages/eslint-plugin/src/rules/no-duplicate-type-constituents.ts:
```diff
@@ -0,0 +1,204 @@
+  meta: {
+    type: 'suggestion',
+    docs: {
+      description: 'Disallow duplicate union/intersection type members',
```
```suggestion
      description: 'Disallow duplicate constituents of union or intersection types.',
```

JoshuaKGoldberg commented on packages/eslint-plugin/docs/rules/no-duplicate-type-constituents.md:
```diff
@@ -0,0 +1,49 @@
+>
+> See **https://typescript-eslint.io/rules/no-duplicate-type-constituents** for documentation.
+
+Although TypeScript supports duplicate union and intersection member values, people usually expect members to have unique values within the same intersection and union. Duplicate values make the code redundant and generally reduce readability.
```
[Docs] Here and elsewhere in the file - we should be consistent with how we phrase it.

```suggestion
Although TypeScript supports duplicate union and intersection constituents, people usually expect members to have unique values within the same intersection and union. Duplicate values make the code redundant and generally reduce readability.
```

JoshuaKGoldberg commented on packages/eslint-plugin/docs/rules/no-duplicate-type-constituents.md:
```diff
@@ -0,0 +1,49 @@
+
+## Rule Details
+
+This rule disallows duplicate union or intersection type members. It only checks duplication on the notation. Members with the same value but different names are not marked duplicates.
```
[Docs] I don't think this is accurate anymore?

I also don't know that people would know what _"same value"_ means. Could you elaborate?
* We do check whether type equality ends up being the same
* But sometimes TypeScript treats things that are the same type as different ones:
    ```ts
    type WrapInValue<T> = { value: T };
    type T = { value: number } | WrapInValue<number>;
    ``` 

I'd suggest adding to the docs some explanation of how this rule determines whether two types are equivalent: that we check whether the syntax is the same, or if TypeScript treats them as the exact same type. This would both help explain edge cases to users and help explain why the code is doing _two_ equality checks.

Yet again I am wishing for https://github.com/microsoft/TypeScript/issues/9879 🥲 which would allow us to skip all this logic, and maybe even combine the rule with `no-redundant-type-constituents`...

JoshuaKGoldberg commented on packages/eslint-plugin/src/rules/no-duplicate-type-constituents.ts:
```diff
@@ -0,0 +1,204 @@
+  ) {
+    if (actualNode.constructor !== expectedNode.constructor) {
+      return false;
+    }
```
Indeed, there are a few lines of code not covered by tests. Which means there should be tests covering those lines, or if that's not possible, those lines can be removed.

My other comments in this review have some test suggestions in them. I'm also thinking maybe:
* `number[]` vs `string[]`
* `string[]` vs `string[]`
* `number[][]` vs `string[]`
* `string[][]` vs `string[][]`
* `ClassA<number>` vs `ClassB<number>`
* `ClassA<number>` vs `ClassB<string>`

Are there any others you're thinking of?

JoshuaKGoldberg commented on packages/eslint-plugin/src/rules/no-duplicate-type-constituents.ts:
```diff
@@ -0,0 +1,204 @@
+    }
+    if (!isRecordType(actualNode) || !isRecordType(expectedNode)) {
+      return false;
+    }
```
This line looks like it's just here for type system coverage. I'm guessing you'll have to use some `as` assertions later on instead of it.

JoshuaKGoldberg reviewed (changes requested):
Getting there, this is looking great! No more structural requests from me on this one 😄 - just docs, testing, and maybe an edge case.

Since this one is getting pretty involved with the type system, if another maintainer (@bradzacher?) who's worked with that a bunch could take a look, that'd be nice.

sajikix commented:
Sorry for the late reply...
And thank you for your kind review.

I made the following changes based on the review.

1. fix rule description and document contents
2. SameAstNode logic
3. show which the previous type is duplicating

### fix rule description and document contents
I have modified the rule description and Document to match the implementation.
Specifically, the following changes were made
- Unified with the word `constituents`
  - [x] https://github.com/typescript-eslint/typescript-eslint/pull/5728#discussion_r1027304702
  - [x] https://github.com/typescript-eslint/typescript-eslint/pull/5728#discussion_r1027304633
- Changed the content of Rule Details
  - Specified that It is comparing Nodes in two different ways.
  - [x] https://github.com/typescript-eslint/typescript-eslint/pull/5728#discussion_r1027305135
- Modified sample case
  - Included examples of both AST Node and Type comparisons.

### SameAstNode logic
As you reviewed, there was some unnecessary processing and test missing in the ASTNode comparison.
- The comparison between `constructor` properties was removed because it did not seem to be used.
  - [x] https://github.com/typescript-eslint/typescript-eslint/pull/5728#discussion_r1027305769
- The `Type guard`, which was confirmed to be an object, has also been replaced with `as`.
  - [x] https://github.com/typescript-eslint/typescript-eslint/pull/5728#discussion_r1027305996
- In addition, I added a test case because there was no test that went through the following process.(Coverage is now 100%.) ([code link](https://github.com/sajikix/typescript-eslint/blob/22c60f859c5e8f1df1a0705a3c333875dd29b0c1/packages/eslint-plugin/src/rules/no-duplicate-type-constituents.ts#L45-L52))
```ts
     if (
      actualNodeKeys.some(
        actualNodeKey =>
          !Object.prototype.hasOwnProperty.call(expectedNode, actualNodeKey),
      )
    ) {
      return false;
    }
```

### show which the previous type is duplicating
Especially when duplicates are detected with a Type match, It might be difficult to know which previous type is duplicating the type.

So I tried to add to the message which type duplicated the previous one like this.([code link](https://github.com/sajikix/typescript-eslint/blob/22c60f859c5e8f1df1a0705a3c333875dd29b0c1/packages/eslint-plugin/src/rules/no-duplicate-type-constituents.ts#L80-L83))
```js
messages: {
    duplicate:
      '{{type}} type member {{duplicated}} is duplicated with {{previous}}.',
},
```

As shown below, when registering a duplicate type, the previous duplicate type is registered as a set. ([code link](https://github.com/sajikix/typescript-eslint/blob/22c60f859c5e8f1df1a0705a3c333875dd29b0c1/packages/eslint-plugin/src/rules/no-duplicate-type-constituents.ts#L126-L135))
```ts
     const duplicatePreviousConstituent = uniqConstituents.find(
          ele => isSameAstNode(ele.node, constituentNode) || ele.type === type,
        );
        if (duplicatePreviousConstituent) {
          duplicateConstituents.push({
            duplicated: constituentNode,
            duplicatePrevious: duplicatePreviousConstituent.node,
          });
          return uniqConstituents;
        }
```

However, I am not confident that this format is easy to understand, so if you have a better method or wording, please let me know.

JoshuaKGoldberg commented on packages/eslint-plugin/src/rules/no-duplicate-type-constituents.ts:
```diff
@@ -0,0 +1,201 @@
+    fixable: 'code',
+    messages: {
+      duplicate:
+        '{{type}} type member {{duplicated}} is duplicated with {{previous}}.',
```
```suggestion
        '{{type}} type constituent {{duplicated}} is duplicated with {{previous}}.',
```

sajikix replied:
✅ done 🙏

JoshuaKGoldberg commented on packages/eslint-plugin/docs/rules/no-duplicate-type-constituents.md:
```diff
@@ -0,0 +1,64 @@
+
+### `ignoreIntersections`
+
+When set to true, duplicate checks on intersection type members are ignored.
```
```suggestion
When set to true, duplicate checks on intersection type constituents are ignored.
```

sajikix replied:
✅ done 🙏

JoshuaKGoldberg commented on packages/eslint-plugin/docs/rules/no-duplicate-type-constituents.md:
```diff
@@ -0,0 +1,64 @@
+
+### `ignoreUnions`
+
+When set to true, duplicate checks on union type members are ignored.
```
```suggestion
When set to true, duplicate checks on union type constituents are ignored.
```

sajikix replied:
✅ done 🙏

JoshuaKGoldberg commented on packages/eslint-plugin/docs/rules/no-duplicate-type-constituents.md:
```diff
@@ -0,0 +1,64 @@
+## Rule Details
+
+This rule disallows duplicate union or intersection constituents.
+
+It determines whether two types are equivalent in the following way.
+
+1. whether the syntax is exactly the same.
+2. whether TypeScript treats them as the same type.
+
+If either of the two conditions is satisfied, It treats the two types as duplicates.
```
I don't think most users really care about how we determine the equivalency internally. The syntax being exactly the same is an internal optimization to avoid using the type checker if we don't have to.

```suggestion
We consider types to be duplicate if they evaluate to the same result in the type system.
For example, given `type A = string` and `type T = string | A`, this rule would flag `string | A`.
```

sajikix replied:
Certainly, there was no need to even talk about internal optimization here...

JoshuaKGoldberg commented on packages/eslint-plugin/tests/rules/no-duplicate-type-constituents.test.ts:
```diff
@@ -0,0 +1,647 @@
+    {
+      code: noFormat`type A = "A";
+type T = (A) ${operator} (A);`,
+      output: noFormat`type A = "A";
```
As of #5379 we don't need to `noFormat` `output`:

```suggestion
      output: `type A = "A";
```

Here and elsewhere.

Also, I don't think we need to `noFormat` this test case's `code` either? Here and elsewhere, I think you can remove a bunch (maybe all) of the `noFormat`s.

sajikix replied:
I didn't know this improvements( #5379 ) were in. Thank you!

JoshuaKGoldberg commented on packages/eslint-plugin/docs/rules/no-duplicate-type-constituents.md:
```diff
@@ -0,0 +1,64 @@
+
+type T3 = { a: string } & { b: string };
+
+type T4 = [1, 2, 3] & [1, 2, 3, 4];
```
This evaluates to `never`, so I'd suggest a `union` type instead:

```suggestion
type T4 = [1, 2, 3] | [1, 2, 3, 4];
```

(and probably the _Incorrect_ code case should be a `|` union too)

sajikix replied:
✅ done 🙏

JoshuaKGoldberg commented on packages/eslint-plugin/docs/rules/no-duplicate-type-constituents.md:
```diff
@@ -0,0 +1,64 @@
+>
+> See **https://typescript-eslint.io/rules/no-duplicate-type-constituents** for documentation.
+
+Although TypeScript supports duplicate union and intersection constituents, people usually expect members to have unique values within the same intersection and union. Duplicate values make the code redundant and generally reduce readability.
```
Just proofreading a bit to make it more clear what a "constituent" is for someone who hasn't heard the term before, like how https://typescript-eslint.io/rules/no-redundant-type-constituents does. How does this make you feel?

```suggestion
Although TypeScript supports types ("constituents") in union and intersection types being duplicates of each other, developers typically expect each constituent to be unique within the same intersection and union.
Duplicate values make the code overly verbose and generally reduce readability.
```

sajikix replied:
Looks great! 
In fact, I was not familiar with the word `constituents` until I started working on this PR, but I think this sentence is much easier to understand.

JoshuaKGoldberg replied:
Haha yeah I learned a ton of random new words working on linting tools. It's why I recommend them to people as a great type of open source project to dive into!

This PR is shaping up nicely btw. I'm really excited to see it in. Awesome work @sajikix! 👏

JoshuaKGoldberg commented on packages/eslint-plugin/src/rules/no-duplicate-type-constituents.ts:
```diff
@@ -0,0 +1,201 @@
+      >((uniqConstituents, constituentNode) => {
+        const type = checker.getTypeAtLocation(
+          parserServices.esTreeNodeToTSNodeMap.get(constituentNode),
+        );
```
The point of the AST comparisons is so that we don't have to call `checker.getTypeAtLocation` if we don't need. But now this code is always checking that API even before it compares the ASTs! We're going to need to make these `type`s _lazily_ computed.

* The code should perform the `isSameAstNode` check _before_ it calls to `checker.getTypeAtLocation`
* If no nodes fail `isSameAstNode` (e.g. `type T = A | A`) then `checker.getTypeAtLocation` should never be called
* `checker.getTypeAtLocation` can be slow, and poking through https://github.com/microsoft/TypeScript/blob/f576398194bc1c16da7cbf4c6bd23f3bf0f41eea/src/compiler/checker.ts#L45148 I don't believe it _fully_ caches its results (@bradzacher has previously informed me on its quirks?). 
    * You might want to make a `Map<TSESTree.Node, Type>` to cache types?

sajikix replied:
I have taken your suggestion and changed the implementation as follows
- First compare values with `isSameAstNode` and do early return if duplicates are found
- Only if no duplicates are found in `isSameAstNode`, call `checker.getTypeAtLocation` to compare type objects
- Type objects are cached as `Map<TSESTree.Node, Type>`

JoshuaKGoldberg commented on packages/eslint-plugin/src/rules/no-duplicate-type-constituents.ts:
```diff
@@ -0,0 +1,201 @@
+
+      const fix: TSESLint.ReportFixFunction = fixer => {
+        return duplicateConstituents
+          .map(duplicateConstituent => {
```
Nit: this way it doesn't need to `.flat()` later ✨ 

```suggestion
          .flatMap(duplicateConstituent => {
```

sajikix replied:
You're absolutely right! (Why didn't I think of that? .... 🤦‍♂️)

JoshuaKGoldberg commented on packages/eslint-plugin/src/rules/no-duplicate-type-constituents.ts:
```diff
@@ -0,0 +1,201 @@
+      const fix: TSESLint.ReportFixFunction = fixer => {
+        return duplicateConstituents
+          .map(duplicateConstituent => {
+            const fixes: TSESLint.RuleFix[] = [];
```
Nit: this `fixes` variable isn't really necessary. The code could be simplified later a bit to return the result of a `.map` directly:

```ts
if (!beforeUnionOrIntersectionToken) {
  return [];
}

// ...

return [ /* ... */ ].map(token => fixer.remove(token));
```

How does that make you feel?

_(I tried this locally to confirm, yes, it does still pass all the tests)_

sajikix replied:
As you suggested, I Implemented to return the result of a `.map` directly 👍

JoshuaKGoldberg commented on packages/eslint-plugin/src/rules/no-duplicate-type-constituents.ts:
```diff
@@ -0,0 +1,201 @@
+      }, []);
+
+      const fix: TSESLint.ReportFixFunction = fixer => {
+        return duplicateConstituents
+          .map(duplicateConstituent => {
```
I'm confused about why this refers to `duplicateConstituents` at all in its fixer function. Each duplicate type node should be able to just fix itself. No need to have each of them provide a fixer for the other nodes.

I changed the code locally to remove `duplicateConstituents.map(duplicateConstituent => ...`, and instead have each node only remove itself. And all the tests pass!

```ts
context.report({
  // ...
  node,
  fix: fixer => {
    const beforeTokens = sourceCode.getTokensBefore(
    // ...
});
```

...and because the `fix` can be inline, there's no need to have a `duplicateConstituents` array at all! `context.report` can be called directly when a duplicate constituent is detected.

sajikix replied:
Indeed, the `fix` function was redundant, and there was no need to loop through the `duplicateConstituents` twice...
Changed as suggested. 👍

JoshuaKGoldberg reviewed (changes requested):
Making progress! 🚀 

Btw, you don't need to give a full summary of all the changes you made each time a review happens. You can just push commits & re-request review, and I'll figure things out 😄. If there's anything you want to call out in particular, it's easiest to just respond in a comment thread.

Alonski commented:
Hey,
Wanted to chime in that I think this is awesome and thank you so much for working on this!

armano2 commented on packages/eslint-plugin/src/rules/no-duplicate-type-constituents.ts:
```diff
@@ -0,0 +1,203 @@
+              beforeUnionOrIntersectionToken,
+              duplicateConstituent.duplicated,
+            );
+            const bracketAfterTokens = afterTokens.slice(
+              0,
+              bracketBeforeTokens.length,
+            );
```
```suggestion
            const bracketAfterTokens = sourceCode.getTokensAfter(
              duplicateConstituent.duplicated,
              { count: bracketBeforeTokens.length },
            );
```

sajikix replied:
Looks very good! 
It was very helpful to learn that there are such useful options 🙏

armano2 commented on packages/eslint-plugin/src/rules/no-duplicate-type-constituents.ts:
```diff
@@ -0,0 +1,203 @@
+            const afterTokens = sourceCode.getTokensAfter(
+              duplicateConstituent.duplicated,
+            );
+            const beforeUnionOrIntersectionToken = beforeTokens
+              .reverse()
+              .find(token => token.value === '|' || token.value === '&');
```
i think this is a little faster
```suggestion
            const beforeTokens = sourceCode.getTokensBefore(
              duplicateConstituent.duplicated,
              { filter: token => token.value === '|' || token.value === '&' },
            );
            const beforeUnionOrIntersectionToken = beforeTokens[beforeTokens.length - 1];
```

sajikix replied:
The way you taught me is more smart 👍

armano2 commented:
i did some testing in playground and ranges seem to be a little off

[playground](https://deploy-preview-5728--typescript-eslint.netlify.app/play/#ts=4.9.3&sourceType=module&code=C4TwDgpgBAglC8UBEMkG4CwAoUkoCEFl91tdoBhIpC0ncaAEWsbrIagBUiAKOAHwIBKKADIofKIPwjBPKoMZDM9PN0STpsiQOFSJCqEpXkoAMV66Z+zXrmGlYnfutyry9ngDill9vn6xthAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oBNZjGyAQ3yJaRUrTIcmyfJXyxELVBkiJo0DtEjgwAXxB6gA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA)

sajikix commented:
Thanks for testing it!

> i did some testing in playground and ranges seem to be a little off

Let me check if my understanding is correct.

In the following example you pointed out to me, there are two duplicated constituents.

 ```ts
type F = (A | B) | (A | B) | (C | D) & (A | B) | (A | B); 
//                  ^^^^^                         ^^^^^
```

In this case, the fixers try to remove following ranges(written in `└  x  ┘`).
```ts
type F = (A | B) | (A | B) | (C | D) & (A | B) | (A | B);
//               └   x   ┘                     └   x   ┘
```
However, fixers do not delete spaces before and after the duplicated constituents(`(A | B)`), so fixers are going to replace  `| (A | B)`(the range written in `└  x  ┘`) with ` `(single space between `|` and `(A | B)`).

If my understanding of the above is correct, I feel the current behavior in [playgroud](https://deploy-preview-5728--typescript-eslint.netlify.app/play/#ts=4.9.3&sourceType=module&code=C4TwDgpgBAglC8UBEMkG4CwAoUkoCEFl91tdoBhIpC0ncaAEWsbrIagBUiAKOAHwIBKKADIofKIPwjBPKoMZDM9PN0STpsiQOFSJCqEpXkoAMV66Z+zXrmGlYnfutyry9ngDill9vn6xthAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oBNZjGyAQ3yJaRUrTIcmyfJXyxELVBkiJo0DtEjgwAXxB6gA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA) (which you created) is correct. .... 🤔 

Or it's very likely that I'm mistaken about something. 😢

armano2 commented:
reported values are correct, but fixers are not

i simplified this example [playground](https://deploy-preview-5728--typescript-eslint.netlify.app/play/#ts=4.9.3&sourceType=module&code=C4TwDgpgBAYlC8UAUBBKAfKAhAlB5amu+SAwvgCJ4BkB+xmq9OA3ALABQQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oBNZjGyAQ3yJaRUrTIcmyfJXyxELVBkiJo0DtEjgwAXxB6gA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA)

```
type F = (A | B) | (A | B) | (C | D) & (A | B) | (A | B);
```

this example has 4 fixers, and after applying any of first 3 fixers result is same, last one:
- 1 fixer result:
```
type F = (A | B) | (A | B) | (C | D) & (A | B)  ;
```
- 2 fixer result:
```
type F = (A | B) | (A | B) | (C | D) & (A | B)  ;
```
- 3 fixer result:
```
type F = (A | B) | (A | B) | (C | D) & (A | B)  ;
```
- 4 fixer result:
```
type F = (A | B) | (A | B) | (C | D)   | (A | B);
```

sajikix commented:
Thank you so much for your detailed explanation!
And I finally realized the problem with my implementation. 🙏 

The problem with the current behavior is that there are multiple fixers (number of detected duplicates) that remove the exact same duplicate (the one at the end), right?
And this kind of fixer behavior is not very good for eslint's autofix behavior: .... 😭 

Now that I am aware of the problem, I would like to implement it according to the second solution you suggested (= report / fixer only the first duplicates found in constituents).
In addition, I will include a change to correctly specify the `node` and `loc` options in the report function. This will make it easier to see which parts are duplicates and which parts will be removed. (You suggested this too!).

However, I am wondering if the above change (pointing out the first detected duplicate) should be applied to the report function as well, or only to the fixer.
In other words, which of the following is the preferred behavior?

1. report and provide `fixer` only the first duplicate detected
```ts
type F = (A | B) | (A | B) | ((C | D) & (A | B)) | (A | B);
// reported         ^^^^^
// fixer            ^^^^^
```

2. provide a fixer only for the first duplicate detected (report all duplicated)
```ts
type F = (A | B) | (A | B) | (C | D) & (A | B) | (A | B);
// reported         ^^^^^     ^^^^^^^^^^^^^^^     ^^^^^
// fixer            ^^^^^
```

sajikix commented:
I have addressed the issues regarding the range of the fixer function and the clarity of the error reporting scope!

You can see these improvements in action by checking out the [playground](https://deploy-preview-5728--typescript-eslint.netlify.app/play/#ts=4.9.3&sourceType=module&code=C4TwDgpgBAglC8UBEMkG4CwAoUkoCEFl91tdoBhIpC0ncaAEWsbrIagDEiAKOAHwIBKKIL6jhEnlUGMRAMijjB+EWIHDMWIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oBNZjGyAQ3yJaRUrTIcmyfJXyxELVBkiJo0DtEjgwAXxB6gA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA). (It should be fixed properly 🙏 )

Talking about implementation, I have cut the `report` function and changed the implementation to call it each time a duplicate is found.

armano2 reviewed (dismissed):
this look great, thank you for your hard work

JoshuaKGoldberg commented on packages/eslint-plugin/src/rules/no-duplicate-type-constituents.ts:
```diff
@@ -0,0 +1,211 @@
+    },
+    schema: [
+      {
+        type: 'object',
```
Nit: good to restrict the objects so people don't pass in different properties (e.g. typos/misspellings) and get confused why they do nothing.

```suggestion
        additionalProperties: false,
        type: 'object',
```

JoshuaKGoldberg commented on packages/eslint-plugin/src/rules/no-duplicate-type-constituents.ts:
```diff
@@ -0,0 +1,211 @@
+    ): void {
+      const cachedTypeMap: Map<TSESTree.TypeNode, Type> = new Map();
+      node.types.reduce<TSESTree.TypeNode[]>(
+        (uniqConstituents, constituentNode) => {
```
[Nitpick] I'm not a fan of abbreviating names in code. It makes it harder to read & understand because you have to memorize all sorts of common abbreviations - which can be different across people / codebases.

```suggestion
        (uniqueConstituents, constituentNode) => {
```

JoshuaKGoldberg commented on packages/eslint-plugin/src/rules/no-duplicate-type-constituents.ts:
```diff
@@ -0,0 +1,211 @@
+      key => !astIgnoreKeys.includes(key),
+    );
+    const expectedNodeKeys = Object.keys(expectedNode).filter(
+      key => !astIgnoreKeys.includes(key),
```
[Non-actionable comment] In theory we could make `astIgnoreKeys` a `Set` to use an O(1) `.has(key)` instead of the O(N=3) `.includes(key)`. In practice I think an array of size 3 is not so terrible for performance. So I think this is fine. Just posting here for reference!

sajikix replied:
You're right, it's an array of size 3, so it doesn't seem to affect performance. 
However, I modified it to use `Set` because it is a recursively called process, so it should be as fast as possible, and most importantly, I think `Set` is more likely to convey my intentions. 😄

JoshuaKGoldberg commented on packages/eslint-plugin/tests/rules/no-duplicate-type-constituents.test.ts:
```diff
@@ -0,0 +1,657 @@
+  },
+});
+
+const valid = (operator: '|' | '&'): TSESLint.ValidTestCase<Options>[] => [
```
[Cleanup] You know, I've really started to sour on generating tests. It's harder to read and harder to perform static analysis on the strings. Let's just write out the tests explicitly.

One nice thing about this rule is that the logic is the same for intersections and unions. So I don't think you need to duplicate test cases for each of them. Just write a few to make sure that both are accounted for, then any edge case tests for constituent checking logic can be just written once. (and wanting to generate tests for combinations of edge cases is a smell to me in that way!)

sajikix replied:
Understood!
I have rewritten the test in naive form (test generation is not used.). Basically, I am testing with `Union`, but I have included some `Intersection` stuff.

JoshuaKGoldberg reviewed (changes requested):
This is looking really great, nice work @sajikix! 👏 

Requesting changes on performance tuning, non-generated tests, and some small touchups. Awesome! Looking forward this getting shipped!

armano2 commented on packages/eslint-plugin/src/rules/no-duplicate-type-constituents.ts:
```diff
@@ -0,0 +1,211 @@
+          const constituentNodeType = checker.getTypeAtLocation(
+            parserServices.esTreeNodeToTSNodeMap.get(constituentNode),
+          );
+          const duplicatedPreviousConstituentInType = [
+            ...cachedTypeMap.entries(),
+          ].find(([, type]) => {
+            return type === constituentNodeType;
+          })?.[0];
```
are you sure about this one? he is looking trough values of map instead of keys

sajikix replied:
That's right! This is much more concise and less computationally intensive! 👍  (And the comparison in `Map` is almost the same as `===`, so there is no problem! )

JoshuaKGoldberg reviewed (approved):
🔥 this is wonderful, thanks a million @sajikix! Really excited to get this out in front of users. Really appreciate you working with us on it!

I'll also modify https://github.com/typescript-eslint/typescript-eslint/discussions/6014 to indicate we'll add this to the `recommended-type-checked` ruleset. Because IMO it should be a recommended rule. 😁

sajikix commented:
Thank you again and again for your kind review! 🙏 
And we are very happy that you will be adding it to the ruleset for v6! 😁

Do I need to do any additional work on this PR? (Is it OK if I leave the rest to you?)

JoshuaKGoldberg commented:
none needed, was just waiting for the build to pass then got distracted 😄
